forester/src/epoch_manager.rs

```diff
@@ -873,48 +877,55 @@ impl<R: RpcConnection, I: Indexer<R>> EpochManager<R, I> {
         let mut retries = 0;
         loop {
-            if !self
-                .check_eligibility(epoch_info, &work_item.tree_account)
-                .await?
-            {
-                debug!("Forester not eligible for this slot, skipping batch");
-                return Ok(None);
-            }
-
             match self
-                .process_transaction_batch(
-                    epoch_info,
-                    transaction_chunk,
-                    proof_chunk,
-                    indexer_chunk,
-                )
+                .check_eligibility(epoch_info, &work_item.tree_account)
                 .await
             {
-                Ok(signature) => {
-                    debug!(
-                        "Work item {:?} processed successfully. Signature: {:?}",
-                        work_item.queue_item_data.hash, signature
-                    );
-                    self.increment_processed_items_count(epoch_info.epoch.epoch)
-                        .await;
-                    return Ok(Some(signature));
+                Ok(_) => {
+                    match self
+                        .process_transaction_batch(
+                            epoch_info,
+                            transaction_chunk,
+                            proof_chunk,
+                            indexer_chunk,
+                        )
+                        .await
+                    {
+                        Ok(signature) => {
+                            debug!(
+                                "Work item {:?} processed successfully. Signature: {:?}",
+                                work_item.queue_item_data.hash, signature
+                            );
+                            self.increment_processed_items_count(epoch_info.epoch.epoch)
+                                .await;
+                            return Ok(Some(signature));
+                        }
+                        Err(e) => {
+                            if retries >= self.config.max_retries {
+                                error!(
+                                    "Max retries reached for work item {:?}. Error: {:?}",
+                                    work_item.queue_item_data.hash, e
+                                );
+                                return Err(e);
+                            }
+                            let delay = BASE_RETRY_DELAY * 2u32.pow(retries as u32);
+                            let jitter = rand::thread_rng().gen_range(0..=50);
+                            sleep(delay + Duration::from_millis(jitter)).await;
+                            retries += 1;
+                            warn!(
+                                "Retrying work item {:?}. Attempt {}/{}",
+                                work_item.queue_item_data.hash, retries, self.config.max_retries
+                            );
+                        }
+                    }
+                }
+                Err(ForesterError::NotEligible) => {
```

Comment:
Nice, are there other.
Are there other errors that we need to handle?
For example, edge cases changing epochs before start or after it has ended (might be covered haven't checked).

Reply:
Yes, there are still many places in forester where I use Forester::Custom(String) generic errors, I should definitely use less generic errors for the specific cases. It's on my list.